EventFilter/Phase2TrackerRawToDigi/plugins/DTHDAQToFEDRawDataConverter.cc

```diff
@@ -0,0 +1,151 @@
+  }
+
+  std::streamsize fileSize = rawFile.tellg();
+  rawFile.seekg(0, std::ios::beg);
+
+  std::vector<char> buffer(fileSize);
+  if (!rawFile.read(buffer.data(), fileSize)) {
+    throw cms::Exception("FileReadError") << "Could not read input file: " << inputFile;
+  }
+
+  rawFile.close();
+  return buffer;  // Ensure move semantics
+}
+
+void DTHDAQToFEDRawDataConverter::printHex(const std::vector<char>& buffer, size_t length) {
+  std::ostringstream hexOutput;
+  hexOutput << "Raw bitstream (first " << length << " bytes): ";
+  for (size_t i = 0; i < std::min(buffer.size(), length); ++i) {
+    hexOutput << std::hex << std::setw(2) << std::setfill('0')
+              << static_cast<unsigned int>(static_cast<unsigned char>(buffer[i])) << " ";
+  }
+  edm::LogInfo("DTHDAQToFEDRawDataConverter") << hexOutput.str();
+}
+
+void DTHDAQToFEDRawDataConverter::parseAndDumpEventData(const std::vector<char>& buffer) {
+  size_t orbitSize = buffer.size() / orbitCount;
+  for (unsigned int orbitIdx = 0; orbitIdx < orbitCount; ++orbitIdx) {
+    size_t startIdx = orbitIdx * orbitSize;
+    edm::LogInfo("DTHDAQToFEDRawDataConverter") << "Parsing Orbit " << orbitIdx + 1;
+
+    if (buffer.size() - startIdx < orbitHeaderSize) {
+      edm::LogError("DTHDAQToFEDRawDataConverter")
+          << "Insufficient data for Orbit Header in Orbit " << orbitIdx + 1;
+      return;
+    }
+
+    uint8_t markerH = static_cast<uint8_t>(buffer[startIdx++]);
+    uint8_t markerO = static_cast<uint8_t>(buffer[startIdx++]);
+    if (markerH != orbitHeaderMarkerH || markerO != orbitHeaderMarkerO) {
+      edm::LogError("DTHDAQToFEDRawDataConverter")
+          << "Invalid Orbit Header marker in Orbit " << orbitIdx + 1 << ": 0x"
+          << std::hex << static_cast<int>(markerH) << " 0x" << static_cast<int>(markerO) << std::dec;
+      return;
+    }
+    edm::LogInfo("DTHDAQToFEDRawDataConverter")
+        << "Orbit Header Marker: 0x" << std::hex << static_cast<int>(markerH)
+        << " 0x" << static_cast<int>(markerO) << std::dec;
+
+    // Read key data fields
+    auto readLittleEndian = [&](const char* data, size_t size) -> uint64_t {
+      uint64_t value = 0;
+      for (size_t i = 0; i < size; ++i) {
+        value |= (static_cast<uint64_t>(static_cast<unsigned char>(data[i])) << (8 * i));
+      }
+      return value;
+    };
+
+    uint16_t version = static_cast<uint16_t>(readLittleEndian(&buffer[startIdx], orbitVersionSize));
+    startIdx += orbitVersionSize;
+    uint32_t sourceId = static_cast<uint32_t>(readLittleEndian(&buffer[startIdx], sourceIdSize));
+    startIdx += sourceIdSize;
+
+    edm::LogInfo("DTHDAQToFEDRawDataConverter")
+        << "Version: " << version << " Source ID: " << sourceId;
+  }
+}
+
+void DTHDAQToFEDRawDataConverter::produce(edm::Event& event, const edm::EventSetup&) {
+  std::vector<char> buffer;
+  try {
+    buffer = readRawFile(inputFile_);
```

Comment:
If the input file contains more than 1 event, I think this code will just output exactly the same data for all events, since it reopens the input file each time produce() is called.